docs/configuration.md
---
title: Configuration
---

The configuration of the OAuth Static Webserver is split into two main parts:

- Server Configuration to set up the server environment and session management.
- Site Configuration to define the static sites to be served and their respective OIDC protection settings.

## Server Configuration

The server configuration is done with environment variables.

| Field                          | Default                                         | Description                                                           |
|:-------------------------------|-------------------------------------------------|-----------------------------------------------------------------------|
| `LOG_LEVEL`                    | `info`                                          | The logging level like `debug`, `info`, `warn` or `error`             |
| `CONFIG_PATH`                  | `/etc/oauth-static-webserver/config.yaml`       | The path to the configuration file.                                   |
| `HOST_ADDRESS`                 |                                                 |                                                                       |
| `HOST_PORT`                    | `8080`                                          | The http(s) listen port.                                              |
| `HTTP2_MAX_CONCURRENT_STREAMS` | `100`                                           |                                                                       |
| `HTTP2_MAX_READ_FRAME_SIZE`    | `1048576` (1 KiB)                               |                                                                       |
| `HTTP2_IDLE_TIMEOUT`           | `10` (Seconds)                                  |                                                                       |
| `TLS_ENABLED`                  | `false`                                         | Enables the TLS functionality of the server.                          |
| `TLS_HTTP_REDIRECT`            | `true`                                          | Force HTTPS instead of HTTP                                           |
| `TLS_CERT_FILE`                |                                                 | The path to the TLS certificate file.                                 |
| `TLS_KEY_FILE`                 |                                                 | The path to the TLS key file.                                         |
| `TLS_AUTO_TLS`                 | `false`                                         | To use automatic TLS certificate request (Let's Encrypt)              |
| `TLS_AUTO_TLS_CERT_CACHE_DIR`  | Uses a tmp directory, when no path is provided. | The cert cache dir, required to prevent Let's Encrypt rate limiting.  |
| `SESSION_KEY`                  |                                                 | Session Encryption Key, should be a secret.                           |
| `SESSION_STORE_DRIVER`         | `filesystem`                                    | The session storage. Use `redis` to use a Redis DB.                   |
| `SESSION_STORE_DIRECTORY`      |                                                 | Path to the session storage (only required when `filesystem` is used) |
| `SESSION_REDIS_ADDRESS`        |                                                 | The Address of the redis server                                       |
| `SESSION_REDIS_PORT`           | `6379`                                          | The Port of the redis server                                          |
| `SESSION_REDIS_USERNAME`       |                                                 | Username for the authentication                                       |
| `SESSION_REDIS_PASSWORD`       |                                                 | Password for the authentication                                       |
| `SESSION_REDIS_DB`             | `0`                                             | Redis DB Index                                                        |
| `SESSION_REDIS_POOL_SIZE`      | `10`                                            | Connection pool size for the Redis DB                                 |

## Site Configuration

The site configuration is done with a YAML configuration file.

```yaml
oidc:
  # Base URL for callback (and more) - must accessible from the internet
  # often the reverse proxy
  base_url: "http://localhost:8080/"
  providers:
    - id: idp
      config_url: "[WELL-KNOWN-URL]"
      client_id: "[CLIENT_ID]"
      client_secret: "[CLIENT_SECRET_KEY]"
static_pages:
  - id: page1
    dir: "page1"
    url: "/static/page1"
  - id: page2
    dir: "/var/www/page2"
    url: "/static/page2"
    protection:
      provider: idp
      expression: "text.has_prefix(user.email, '@example.com') || user.level >= 3"
      groups:
        - group1
        - group2 
```

The example above shows the basic structure of the configuration file with all existing options.

Some explanations about the fields:

- `oidc.base_url`: The base URL where the server is reachable from the internet. This is often the URL of the reverse proxy in front of the server.
- `oidc.providers`: A list of OIDC providers to use for authentication.
  - `id`: A unique identifier for the provider.
  - `config_url`: The well-known URL of the OIDC provider.
  - `client_id`: The client ID for the OIDC application.
  - `client_secret`: The client secret for the OIDC application.
- `static_pages`: A list of static pages to be served.
  - `id`: A unique identifier for the static page (used for logging).
  - `dir`: The directory where the static content is located.
  - `url`: The URL path where the static content will be accessible.
  - `protection`: (Optional) The protection settings for the static page.
    - `provider`: The `id` of the OIDC provider to use for authentication.
    - `groups`: (Optional) A list of groups that are allowed to access the static page. If not specified, the group check is skipped.
    - `expression`: (Optional) A custom expression to evaluate for access control. The expression can use user attributes like `user.email`, `user.level`, etc.

The `protection` section is optional. If it is not provided, the static page will be publicly accessible without authentication.
Also, both `groups` and `expression` are optional inside the `protection`.
The expression will be evaluated first and then the group check.

The expression language is a fully featured programming language, but here it will be inserted into a boolean context. 
The language named "tengo" is documented here: https://github.com/d5/tengo/

Also for the expression context, the following stdlib modules are available:

- `text`: https://github.com/d5/tengo/blob/master/docs/stdlib-text.md
- `math`: https://github.com/d5/tengo/blob/master/docs/stdlib-math.md
- `times`: https://github.com/d5/tengo/blob/master/docs/stdlib-times.md

To use it, write the name first and then call the methods, e.g. `text.has_prefix(user.email, '@example.com')`.
All user attributes are available via the `user` variable, e.g. `user.email`, `user.name`, `user.level`, etc.


config.go
package main

import (
	"errors"
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/go-playground/validator/v10"
	"github.com/ilyakaznacheev/cleanenv"
	"golang.org/x/net/http2"
	"gopkg.in/yaml.v3"

	log "github.com/sirupsen/logrus"
)

// --- Config type declaration ---

// Config is the main configuration struct containing settings and content config
type Config struct {
	// contains the main settings for the webserver
	Settings Settings
	// contains the configuration for the pages and security
	Content ContentConfig
}

type Settings struct {
	LogLevel   string          `env:"LOG_LEVEL" env-default:"info" env-description:"Logging level: debug, info, warn, error"`
	Host       SettingsHost    `env-prefix:"HOST_"`
	HTTP2      SettingsHTTP2   `env-prefix:"HTTP2_"`
	TLS        SettingsTLS     `env-prefix:"TLS_"`
	Session    SettingsSession `env-prefix:"SESSION_"`
	ConfigPath string          `env:"CONFIG_PATH" env-default:"/etc/oauth-resource-proxy/config.yaml"`
}

type SettingsSession struct {
	Key string `env:"KEY"`
	// Possible values: "filesystem", "redis"
	StoreDriver string `env:"STORE_DRIVER" env-default:"filesystem" env-description:"Session store driver: filesystem or redis"`
	// when redis
	Redis struct {
		Address  string `env:"ADDRESS"`
		Port     int    `env:"PORT" env-default:"6379"`
		Username string `env:"USERNAME"`
		Password string `env:"PASSWORD"`
		DB       int    `env:"DB" env-default:"0"`
		PoolSize int    `env:"POOL_SIZE" env-default:"10"`
	} `env-prefix:"REDIS_"`
	// when filesystem
	StoreDirectory string `env:"STORE_DIRECTORY"`
}

type SettingsHost struct {
	Address string `env:"ADDRESS"`
	Port    int    `env:"PORT" env-default:"8080"`
}

type SettingsHTTP2 struct {
	MaxConcurrentStreams int `env:"MAX_CONCURRENT_STREAMS" env-default:"100"`
	// frame size in bytes
	MaxReadFrameSize int `env:"MAX_READ_FRAME_SIZE" env-default:"1048576"`
	// timeout in seconds
	IdleTimeout int `env:"IDLE_TIMEOUT" env-default:"10"`
}

type SettingsTLS struct {
	Enabled             bool   `env:"ENABLED" env-default:"false"`
	HTTPRedirect        bool   `env:"HTTP_REDIRECT" env-default:"true"`
	CertFile            string `env:"CERT_FILE"`
	KeyFile             string `env:"KEY_FILE"`
	AutoTLS             bool   `env:"AUTO_TLS" env-default:"false"`
	AutoTLSCertCacheDir string `env:"AUTO_TLS_CERT_CACHE_DIR"`
}

type ContentConfig struct {
	OIDC        ContentConfigOIDC `yaml:"oidc" validate:"required"`
	StaticPages []StaticPage      `yaml:"static_pages" validate:"dive,required"`
}

type ContentConfigOIDC struct {
	BaseUrl   string         `yaml:"base_url" validate:"required,url"`
	Providers []OIDCProvider `yaml:"providers" validate:"dive,required"`
}

type OIDCProvider struct {
	Id           string `yaml:"id" validate:"alphanum"`
	ConfigUrl    string `yaml:"config_url" validate:"required,url"`
	ClientID     string `yaml:"client_id" validate:"alphanum"`
	ClientSecret string `yaml:"client_secret" validate:"alphanum"`
}

type StaticPage struct {
	Id         string                `yaml:"id" validate:"alphanum"`
	Dir        string                `yaml:"dir" validate:"dir"`
	Url        string                `yaml:"url" validate:"required,uri"`
	Protection *StaticPageProtection `yaml:"protection"`
}

type StaticPageProtection struct {
	Provider   string   `yaml:"provider" validate:"alphanum"`
	Expression string   `yaml:"expression"`
	Groups     []string `yaml:"groups" validate:"dive,alphanum"`
}

// --- Config loading and processing ---

// loadConfig loads the configuration from environment variables and config file
func loadConfig() (*Config, error) {
	cfg := new(Config)
	// first load settings from env
	settings, err := loadSettingsFromEnv()
	if err != nil {
		help, errHelp := cleanenv.GetDescription(&cfg, nil)
		if errHelp != nil {
			log.WithError(err).WithError(errHelp).Error("can not get help text")
		} else {
			log.WithError(err).Fatal(help)
		}
		return nil, err
	}
	cfg.Settings = settings
	// then load content config from file
	contentCfg, err := loadContentConfig(settings.ConfigPath)
	if err != nil {
		return nil, err
	}
	cfg.Content = *contentCfg
	return cfg, nil
}

// LoadAndProcessConfig loads, validates and resolves the configuration
func LoadAndProcessConfig() (*Config, error) {
	cfg, err := loadConfig()
	if err != nil {
		log.WithError(err).Fatal("error loading config")
	}

	// apply log level
	level, err := log.ParseLevel(cfg.Settings.LogLevel)
	if err != nil {
		log.WithError(err).Fatal("invalid log level in config")
	}
	log.SetLevel(level)

	validate := validator.New(validator.WithRequiredStructEnabled())
	err = cfg.Validate(validate)
	if err != nil {
		log.WithError(err).Fatal("configuration are not valid")
	}
	log.Info("Config read and validated successfully")

	err = cfg.Process()
	if err != nil {
		log.WithError(err).Error("Error resolving config")
		return nil, err
	}
	return cfg, nil
}

func (c *Config) Validate(validate *validator.Validate) error {
	return c.Content.Validate(validate)
}

func (c *Config) Process() error {
	return c.Content.Process()
}

func loadSettingsFromEnv() (Settings, error) {
	var settings Settings
	err := cleanenv.ReadEnv(&settings)
	return settings, err
}

func (s Settings) GetWSAddress() string {
	return fmt.Sprintf("%s:%d", s.Host.Address, s.Host.Port)
}

// GetHttps2Server builds a http2.Server from the settings
func (s SettingsHTTP2) GetHttps2Server() *http2.Server {
	return &http2.Server{
		MaxConcurrentStreams: uint32(s.MaxConcurrentStreams),
		MaxReadFrameSize:     uint32(s.MaxReadFrameSize),
		IdleTimeout:          time.Duration(s.IdleTimeout) * time.Second,
	}
}

func (c *ContentConfig) Validate(validate *validator.Validate) error {
	err := validateStruct(validate, c)
	if err != nil {
		return fmt.Errorf("configuration validation failed: %w", err)
	}

	// check all static page protections reference valid Providers
	for _, staticPage := range c.StaticPages {
		if staticPage.Protection != nil {
			err := validateStruct(validate, staticPage.Protection)
			if err != nil {
				return fmt.Errorf("static page %q protection validation failed: %w", staticPage.Id, err)
			}
		}
	}
	return nil
}

func (c *ContentConfig) Process() error {
	c.OIDC.BaseUrl = strings.TrimRight(c.OIDC.BaseUrl, "/")
	return nil
}

// --- helper and utility functions ---

func validateStruct(validate *validator.Validate, value any) error {
	err := validate.Struct(value)
	if err != nil {
		var invalidValidationError *validator.InvalidValidationError
		if errors.As(err, &invalidValidationError) {
			log.WithError(err).Warn("error validating struct")
			return err
		}
	}
	return err
}

func loadContentConfig(path string) (*ContentConfig, error) {
	var contentCfg ContentConfig
	err := loadConfigFromFile(path, &contentCfg)
	if err != nil {
		return nil, err
	}
	return &contentCfg, nil
}

func loadConfigFromFile(path string, contentCfg *ContentConfig) error {
	file, err := os.ReadFile(path)
	if err != nil {
		return err
	}
	return yaml.Unmarshal(file, contentCfg)
}
